src/ripple/protocol/impl/tokens.cpp

```diff
@@ -149,103 +174,324 @@ encodeBase58(
-static
-std::string
-decodeBase58 (std::string const& s,
-    InverseArray const& inv)
+static boost::optional<std::pair<Slice, DecodeMetadata>>
+decodeBase58(
+    Slice in,
+    MutableSlice out,
+    InverseArray const& inv,
+    DecodeBase58Detail::bitset const& flags)
 {
-    auto psz = s.c_str();
-    auto remain = s.size();
+    using namespace boost::multiprecision;
+
+    // Given an input of size N, the output size is not known exactly. If the largest token
+    // size is MaxDecodedTokenBytes, this function needs to be able to decode MaxDecodedTokenBytes+1
+    // bytes
+    constexpr size_t maxOutBytes = MaxDecodedTokenBytes+1;
+    constexpr size_t maxOutBits = maxOutBytes*8;
+
     // Skip and count leading zeroes
     int zeroes = 0;
-    while (remain > 0 && inv[*psz] == 0)
+    while (in.size() > 0 && inv[in[0]] == 0)
     {
         ++zeroes;
-        ++psz;
-        --remain;
+        in+=1;
+    }
+
+    {
+        std::size_t const maxSize = in.size() * 733 / 1000 + 1;
+        if (out.size() + 1 + rippleLibEncodedSeedPrefix.size() + checksumBytes <
+            maxSize - 1)  // -1 because maxSize may overestimate
+            return {};
+        // tighter constraint for non-ripplelib encoded data
+        // ripple-lib encoded will be <1 token><2 encoding><16 data><4 checksum>==23
+        if (!flags.test(DecodeBase58Detail::maybeRippleLibEncoded) &&
+            out.size() + 1 + checksumBytes < maxSize - 1) // -1 because maxSize may overestimate
+        {
+            return {};
+        }
+        if (maxSize + zeroes < checksumBytes || maxSize + zeroes > maxOutBytes)
+            return {};
     }
-    // Allocate enough space in big-endian base256 representation.
-    // log(58) / log(256), rounded up.
-    std::vector<unsigned char> b256(
-        remain * 733 / 1000 + 1);
-    while (remain > 0)
+
+    if (out.size() > maxOutBytes)
+    {
+        return {};
+    }
+
+    // coefficients for the base 58^10 decoding
+    std::array<std::uint64_t, 6> b5810{};
+    // Durring loop: index of the last next coefficients to populate
+    // After loop: number of coefficients in the b5810 array.
+    int b5810i = 0;
+    {
+        // convert from base58 to base 58^10; All values will fit in a 64-bit
+        // uint without overflow
+
+        // pow58[i] == 58^i
+        constexpr std::array<std::uint64_t, 10> pow58{1,
+                                                      58,
+                                                      3364,
+                                                      195112,
+                                                      11316496,
+                                                      656356768,
+                                                      38068692544,
+                                                      2207984167552,
+                                                      128063081718016,
+                                                      7427658739644928};
```

Comment:
no change needed, but I wonder if something like:
```
    constexpr std::array<std::uint64_t, 10> pow58{
        1ull,
        58ull,
        58ull*58,
        58ull*58*58,
        58ull*58*58*58,
        58ull*58*58*58*58,
        58ull*58*58*58*58*58,
        58ull*58*58*58*58*58*58,
        58ull*58*58*58*58*58*58*58,
        58ull*58*58*58*58*58*58*58*58};
```
would be easier for others to read. Either way, the values are clear, even if you can't do 58^9 in your head.

Reply:
Or, we can just write some code and let the compiler do all the work:

```
constexpr auto pow58 = []()
{
    std::array<std::uint64_t, 10> powtab { 1 };

    for (std::size_t i = 1; i != powtab.size(); ++i)
        powtab[i] = 58 * powtab[i - 1];

    return powtab;
}();
```

Reply:
Lambda aren't constexpr until c++-17, so we can't use that yet.

Reply:
Oh, duh.